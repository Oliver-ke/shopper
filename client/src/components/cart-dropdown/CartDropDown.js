import React from 'react';
import CartItem from '../cart-item/CartItem';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { selectCartItems } from '../../redux/cart/cartSelectors';
import { toggleCartHidden } from '../../redux/cart/cartAction';
import CustomButton from '../custom-button/CustomButton';
import { CartItemContainer, DropdownContainer, Message } from './cartDropDownStyle';

const CartDropDown = ({ cartItems, history, dispatch }) => {
	const handleCheckout = () => {
		dispatch(toggleCartHidden());
		return history.push('/checkout');
	};
	return (
		<DropdownContainer>
			<CartItemContainer>
				{cartItems.length ? (
					cartItems.map((item) => <CartItem key={item.id} item={item} />)
				) : (
					<Message>No Item in cart</Message>
				)}
			</CartItemContainer>
			<CustomButton onClick={handleCheckout}>Go To Checkout</CustomButton>
		</DropdownContainer>
	);
};

// const mapStateToProps = (state) => ({
// 	cartItems: selectCartItems(state),
// });

const mapStateToProps = createStructuredSelector({
	cartItems: selectCartItems,
});
export default connect(mapStateToProps)(withRouter(CartDropDown));

// or
// export default withRouter(connect(mapStateToProps)(CartDropDown));
// use goat tie rope, use rope tie goat, thay're all still the same thing
