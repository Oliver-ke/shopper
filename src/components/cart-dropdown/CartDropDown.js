import React from 'react';
import CartItem from '../cart-item/CartItem';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { selectCartItems } from '../../redux/cart/cartSelectors';
import CustomButton from '../custom-button/CustomButton';
import './cartDropDown.scss';

const CartDropDown = ({ cartItems }) => {
	return (
		<div className="cart-dropdown">
			<div className="cart-items">{cartItems.map((item) => <CartItem key={item.id} item={item} />)}</div>
			<CustomButton>Go To Checkout</CustomButton>
		</div>
	);
};

// const mapStateToProps = (state) => ({
// 	cartItems: selectCartItems(state),
// });

const mapStateToProps = createStructuredSelector({
	cartItems: selectCartItems,
});
export default connect(mapStateToProps)(CartDropDown);
