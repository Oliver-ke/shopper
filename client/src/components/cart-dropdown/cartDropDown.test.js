import React from 'react';
import { shallow } from 'enzyme';
import { CartDropDown } from './CartDropDown';

describe('CartDropDown component', () => {
	let wrapper;
	let mockDispatch;
	let mockProps;
	beforeEach(() => {
		mockDispatch = jest.fn();
		mockProps = {
			dispatch: mockDispatch,
			cartItems: [ { id: 1, name: 'some name' }, { id: 2, name: 'another name' } ],
			history: {
				push: jest.fn(),
			},
		};
		wrapper = shallow(<CartDropDown {...mockProps} />);
	});

	it('should render CartDropDown component', () => {
		expect(wrapper).toMatchSnapshot();
	});
	it('should call dispatch when clicked', () => {
		wrapper.find('CartDropdownButton').simulate('click');
		expect(mockDispatch).toHaveBeenCalled();
		expect(mockProps.history.push).toHaveBeenCalledWith('/checkout');
	});
});

// cartItems, history, dispatch
