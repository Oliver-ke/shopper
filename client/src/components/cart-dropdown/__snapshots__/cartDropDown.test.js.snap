// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropDown component should render CartDropDown component 1`] = `ShallowWrapper {}`;
